File: rails/app/models/stream.rb
class Stream < ApplicationRecord
  self.table_name = 'streams'

  enum :status, { scheduled: 0, live: 1, ended: 2 }, prefix: :status
  enum :platform, { twitch: 0, you_tube: 1, kick_stream: 2, platform: 3 }, prefix: :platform
  enum :language, { e_n: 0, d_e: 1, f_r: 2, i_t: 3, e_s: 4, j_p: 5, p_t: 6 }, prefix: :language

  belongs_to :tournament, class_name: 'Tournament', inverse_of: :streams, optional: true
  belongs_to :streamer, class_name: 'Player', inverse_of: :streams

  validates :title, presence: true, length: { maximum: 300 }
  validates :stream_url, presence: true, length: { maximum: 200 }

  # Domain invariants — simple rules
  validate :validate_rules

  def validate_rules
    errors.add(:viewer_count_not_negative, 'Peak viewer count must not be negative') unless ((viewer_count_peak.nil? || viewer_count_peak >= 0))
  end

  # Domain invariants — IMPLIES rules
  validate :validate_implies

  def validate_implies
    errors.add(:base, 'actual_start_requires_live_or_ended') if (!actual_start.nil?) && !(status == 'live')
    errors.add(:base, 'ended_at can only be set when stream status is Ended') if (!ended_at.nil?) && !(status == 'ended')
  end

  def to_s
    title.to_s
  end

  # Business operations

  def go_live
    # TODO: implement go_live
  end

  def end
    # TODO: implement end
  end

  def update_viewer_peak(count)
    # TODO: implement update_viewer_peak
  end

  def duration_minutes
    # TODO: implement duration_minutes
    nil
  end

  # Lifecycle state machine
  ALLOWED_TRANSITIONS = {
    'scheduled' => ['live'],
    'live' => ['ended'],
  }.freeze

  def assert_transition!(to_state)
    allowed = ALLOWED_TRANSITIONS.fetch(status.to_s, [])
    unless allowed.include?(to_state.to_s)
      raise ArgumentError, "Transition #{status} -> #{to_state} not allowed"
    end
  end

  def as_json(options = {})
    hash = super(options)
    hash['scheduledStart'] = hash.delete('scheduled_start') if hash.key?('scheduled_start')
    hash['actualStart'] = hash.delete('actual_start') if hash.key?('actual_start')
    hash['endedAt'] = hash.delete('ended_at') if hash.key?('ended_at')
    hash
  end
end

File: laravel/app/Models/Content/Stream.php
<?php

namespace App\Models\Content;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use App\Models\Tournaments\Tournament;
use App\Models\Players\Player;

class Stream extends Model
{
    protected $table = 'streams';

    protected $fillable = ['title', 'stream_url', 'platform', 'status', 'viewer_count_peak', 'scheduled_start', 'actual_start', 'ended_at', 'vod_url', 'tournament_id', 'streamer_id'];

    protected $casts = [
        'scheduled_start' => 'datetime',
        'actual_start' => 'datetime',
        'ended_at' => 'datetime',
    ];

    const PLATFORM_VALUES = ['Twitch', 'YouTube', 'KickStream', 'Platform'];
    const STATUS_VALUES = ['Scheduled', 'Live', 'Ended'];

    public function tournament(): BelongsTo
    {
        return $this->belongsTo(Tournament::class, 'tournament_id');
    }

    public function streamer(): BelongsTo
    {
        return $this->belongsTo(Player::class, 'streamer_id');
    }

    // ── Domain invariants (IMPLIES rules) ───────────────────────────────
    public function validateImplies(): void
    {
        if ($this->actual_start !== null && !($this->status === 'Live')) {
            throw new \RuntimeException('actual_start_requires_live_or_ended');
        }
    }

    // ── Business operations ──────────────────────────────────────────

    public function goLive(): void
    {
        throw new \RuntimeException('go_live not implemented');
    }

    public function end(): void
    {
        throw new \RuntimeException('end not implemented');
    }

    public function updateViewerPeak($count): void
    {
        throw new \RuntimeException('update_viewer_peak not implemented');
    }

    public function durationMinutes(): int
    {
        throw new \RuntimeException('duration_minutes not implemented');
    }

}
